proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,153 @@
+# RPC design refactoring
+
+* Proposal: [SDL-NNNN](nnnn-rpc-design-refactoring.md)
+* Author: [Alexander Kutsan](https://github.com/LuxoftAKutsan)
+* Status: **Awaiting review**
+* Impacted Platforms: [Core]
+
+## Introduction
+
+This proposal is to rework design of application manager layer in order to:
+ - Reduce coupling between components
+ - Reduce responsibility of ApplicationManager class
+ - Provide plugin architecture for easy adding additional functionality and RPC groups   
+
+## Motivation
+
+SDL core project is actively implementing new features and new functionality.  
+To keep project on track and to reduce risks of regression during implementation
+refactoring of application manager layer should be done.  
+
+Currently, extending SDL functionality may require code duplication. 
+RC module, in fact, duplicates logic of application_manager layer, there is separate request controller, separate mechanism to call  polcies checks, and RPC's processing. So any common logic that should be implemented in application_manager layer should be duplicated in request controller, it results in bugs, and requires more time for implementation. 
+
+The following steps will provide ability to keep RC enough encapsulated, without duplicating SDL code :
+
+## Proposed solution
+#### Reduce coupling between components and use dependency injection practice
+
+Now Application manager is a service locator and should to be passed to each component
+even if component does not need application manager, but need some other classes that owned by application manager.
+During construction components should explicitly receive all dependencies. 
+It will prevent from unwanted access and improve encapsulation. 
```

Comment:
It will prevent from unwanted access and improve encapsulation. 
should become 
This practice will improve encapsulation for the project and make code much more maintainable going forward.

Reply:
@mrapitis, done in 2c9f7a3f7bb7e6c8b26bb1e4aec779cf65733666